### PR TITLE
WIP grafana/ui: namespaced exports

### PR DESCRIPTION
**Pull request by dprokop**
Ref #20151

This introduces reorg of grafana/ui to enable namespaced imports. Currently grafana/ui already supports that (user can do i.e `import XXX from '@grafanaui/components/XXX'`. But all of the components live in the `components` directory and are not organised anyhow.

There is no TS suggested way of doing submodules, so the solution is to have index.ts files in the namespaced directories, which export the namespace related code. See https://github.com/microsoft/TypeScript/issues/8305

As a first stab I'm moving the visualizations (Graph, Gauge, BarGauge etc...) into top level `visualzations` dir, which enables importing like this: `import { Graph } from '@grafana/ui/visualizations';` 

Following namespaces could be:
- `general` - i.e. Drawer, Cascader,  Modal, ColorPicker, Icon, etc
- `forms` - everything that's form's related, meaning all new forms components which now are imported with `import { Form } from '@grafana/ui'`
- `charts` - Legend, Tooltip
- `navigation` - Tabs
- `data` - transformations, datalinks
- `panel` - thresholds editor, value mappings editor, panel options components, unit picker etc

TODO:
- [ ] Come up with namespaces
- [ ] Move components accordingly

### CODE REVIEW

**Comment by dprokop** [comment]
@ryantxu @torkelo not sure about the proposed namespaces above, any other ideas?

**Comment by torkelo** [comment]
the only thing annoying that I could not find a way around (when I looked at this in the beginning) is that when importing from the app we need to import from the "src" dir. So someone could not copy imports from the grafana core app to a plugin. But think that is fine, the only fix for that is removing src dir and moving everything to root or isolating the webpack build of @grafana/ packages form the main app.

**Comment by torkelo** [comment]
> Following namespaces could be:
general - i.e. Drawer, Cascader, Modal, ColorPicker, Icon, etc
forms - everything that's form's related, meaning all new forms components which now are imported with import { Form } from '@grafana/ui'
charts - Legend, Tooltip
navigation - Tabs
data - transformations, datalinks
panel - thresholds editor, value mappings editor, panel options components, unit picker etc

For data (transformation UI, data links etc), think that could go under panel/  to not confuse it with @grafana/data.  Wonder what more navigation components we will have or if that can live in general as well.

**Comment by dprokop** [comment]
@torkelo ref:

> the only thing annoying that I could not find a way around (when I looked at this in the beginning) is that when importing from the app we need to import from the "src" dir. 

see ecf13cfd090e1f09145ba70b617104604750c58c . I have added paths mapping in core's tsconfig.json file so that any import from `@grafana/ui/*` is mapped to grafana-ui's `src directory. This makes it possible to use namespaced imports in core. But specifying the path tsconfig is not enough, as webpack does not understand that out of the box. We need to specify aliases for webpack to pick up the correct  paths. To not do it manually I have added https://github.com/dividab/tsconfig-paths-webpack-plugin which does that automatically. 

WDYT?

**Comment by torkelo** [comment]
looks good!

**Comment by dprokop** [comment]
> Wonder what more navigation components we will have or if that can live in general as well

Since the Tabs will be the only "navigation" component for now I think we can start with general namespace for now. I'm not super wild about the `general` name though, maybe `components` should to the job. Ideas?

**Comment by mckn** [comment]
> > Wonder what more navigation components we will have or if that can live in general as well
> 
> Since the Tabs will be the only "navigation" component for now I think we can start with general namespace for now. I'm not super wild about the `general` name though, maybe `components` should to the job. Ideas?

Would `core` be a good name?

**Comment by dprokop** [comment]
panel / data -> `editors`

**Comment by dprokop** [comment]
@ryantxu we agreed with @torkelo to keep the core imports to namespaces as `**/src/[namespace]` as adding mapping to enable `**/[namespace]` is very problematic.

File: public/app/features/plugins/plugin_loader.ts [hunk 1]
```diff
@@ -43,17 +43,28 @@ import * as d3 from 'd3';
...
 grafanaUI.DataSourcePlugin = grafanaData.DataSourcePlugin;
 grafanaUI.AppPlugin = grafanaData.AppPlugin;
 grafanaUI.DataSourceApi = grafanaData.DataSourceApi;
 
+/**
+ * 6.XXX introduces namespaced exports from @grafana/ui
+ * This makes sure that exports previously accessible via direct imports from @grafana/ui are still available
+ * via global namespace
+ */
+grafanaUI = {
```
Comment (dprokop) [inline public/app/features/plugins/plugin_loader.ts]: @torkelo @ryantxu adding namespaced exports to global grafana/ui namespace for backward compatibility

**Comment by dprokop** [comment]
ah shit, for some reason the namespaced imports don't work in plugins :/

**Comment by ryantxu** [comment]
Do we need to add them to the 'exposeToPlugins' list?

**Comment by dprokop** [comment]
@ryantxu no, not really. The thing is webpack have problems building the plugin. I mean it builds but at the same time throws an error that it cannot resolve that path:

image

It tries to resolve it in plugin's `src` dir even though the resolve paths are:

```
modules: [path.resolve(process.cwd(), 'src'), 'node_modules'],
```

@stevenvachon maybe you have some ideas around this?

**Comment by dprokop** [comment]
Btw you guys can try testing this work out with @grafana/ui@test, published last commit under `test` dist tag https://www.npmjs.com/package/@grafana/ui (6.6.0-pre-f27e03c)

**Comment by dprokop** [comment]
@ryantxu but just now I have realised, that the imports won't work anyways... We are not bundling packages with the plugins so if the plugin is imported via `@grafana/ui/namespace` path it's gonna throw as internal paths are with src in them... Not really sure what to do about this

**Comment by torkelo** [comment]
Ah, dammit , think we need to export all the sub paths via systemjs as well

**Comment by dprokop** [comment]
Yeah. This leads me to a thoughts that maybe we should take the simple way and treat the namespaces as exports, as we currently do with Forms. `import { Namespace } from ...`

**Comment by dprokop** [comment]
Ref https://github.com/grafana/grafana/pull/21326#issuecomment-573087996 solved -> we have @grafana packages specified as external dependencies in the webpack config literally (@grafana/ui,  @grafana/data, etc). Changing tem to regex makes namespaced imports possible in the plugins, seems like tests are passing as well. Maybe this battle is not lost yet ;) Will continue some work around this. cc @ryantxu @torkelo

**Comment by dprokop** [comment]
@torkelo you were right about namespaces being necessary to be exposed to plugins via SystemJS. I've just tested it against https://github.com/dprokop/grafana-plugin-namespacing with 01b2860 commit and it seems like it's working.

How do you feel about the need for specifying the exposed namespaces manually as here https://github.com/grafana/grafana/pull/21326/commits/01b28607a11648153d85d438b3fe96b79b869e3e ?

**Comment by torkelo** [comment]
let's chat about this tomorrow, I am a bit on the fence on a few things

**Pull request closed without merging**